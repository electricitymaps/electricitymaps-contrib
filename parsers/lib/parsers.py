import importlib

from electricitymap.contrib.config import EXCHANGES_CONFIG, ZONES_CONFIG

# Prepare all parsers
CONSUMPTION_PARSERS = {}
PRODUCTION_PARSERS = {}
PRODUCTION_PER_MODE_FORECAST_PARSERS = {}
PRODUCTION_PER_UNIT_PARSERS = {}
EXCHANGE_PARSERS = {}
PRICE_PARSERS = {}
CONSUMPTION_FORECAST_PARSERS = {}
GENERATION_FORECAST_PARSERS = {}
EXCHANGE_FORECAST_PARSERS = {}
PRODUCTION_CAPACITY_PARSERS = {}
DAYAHEAD_LOCATIONAL_MARGINAL_PRICE_PARSERS = {}
REALTIME_LOCATIONAL_MARGINAL_PRICE_PARSERS = {}

PARSER_KEY_TO_DICT = {
    "consumption": CONSUMPTION_PARSERS,
    "production": PRODUCTION_PARSERS,
    "productionPerUnit": PRODUCTION_PER_UNIT_PARSERS,
    "productionPerModeForecast": PRODUCTION_PER_MODE_FORECAST_PARSERS,
    "exchange": EXCHANGE_PARSERS,
    "price": PRICE_PARSERS,
    "dayaheadLocationalMarginalPrice": DAYAHEAD_LOCATIONAL_MARGINAL_PRICE_PARSERS,
    "realtimeLocationalMarginalPrice": REALTIME_LOCATIONAL_MARGINAL_PRICE_PARSERS,
    "consumptionForecast": CONSUMPTION_FORECAST_PARSERS,
    "generationForecast": GENERATION_FORECAST_PARSERS,
    "exchangeForecast": EXCHANGE_FORECAST_PARSERS,
    "productionCapacity": PRODUCTION_CAPACITY_PARSERS,
}


def _parser_key_to_parser_folder(parser_key: str):
    return (
        "electricitymap.contrib.capacity_parsers"
        if parser_key == "productionCapacity"
        else "parsers"
    )


# Read all zones
for zone_id, zone_config in ZONES_CONFIG.items():
    for parser_key, v in zone_config.get("parsers", {}).items():
        mod_name, fun_name = v.split(".")
        mod = importlib.import_module(
            f"{_parser_key_to_parser_folder(parser_key)}.{mod_name}"
        )
        PARSER_KEY_TO_DICT[parser_key][zone_id] = getattr(mod, fun_name)


# Read all exchanges
for exchange_id, exchange_config in EXCHANGES_CONFIG.items():
    for parser_key, v in exchange_config.get("parsers", {}).items():
        mod_name, fun_name = v.split(".")
        mod = importlib.import_module(f"parsers.{mod_name}")
        PARSER_KEY_TO_DICT[parser_key][exchange_id] = getattr(mod, fun_name)
