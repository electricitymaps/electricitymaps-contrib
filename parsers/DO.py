#!/usr/bin/env python3

from datetime import datetime, timedelta
from logging import Logger, getLogger
from zoneinfo import ZoneInfo

import pandas as pd
from bs4 import BeautifulSoup
from requests import Session

from electricitymap.contrib.lib.models.event_lists import ProductionBreakdownList
from electricitymap.contrib.lib.models.events import ProductionMix
from electricitymap.contrib.lib.types import ZoneKey
from parsers.lib.config import use_proxy

# This parser gets hourly electricity generation data from oc.org.do for the Dominican Republic.
# The data is in MWh but since it is updated hourly we can view it as MW.
# Solar generation now has some data available but multiple projects are planned/under construction.

DO_SOURCE = "oc.org.do"
URL = "https://apps.oc.org.do/reportesgraficos/reportepostdespacho.aspx"

TOTAL_RENEWABLES_MAPPING = {
    "Total E\xf3lico": "wind",
    "Total Hidroel\xe9ctrica": "hydro",
    "Total Solar": "solar",
}

TOTAL_MAPPING = {
    "Total T\xe9rmico": "Thermal",
    "Total Generado": "Generated",
}

# Power plant types
# http://www.sie.gob.do/images/Estadisticas/MEM/GeneracionDiariaEnero2017/
# Reporte_diario_de_generacion_31_enero_2017_merged2.pdf

THERMAL_PLANTS = {
    "AES ANDRES": "gas",
    "BARAHONA CARBON": "coal",
    "BERSAL": "oil",
    "CEPP 1": "oil",
    "CEPP 2": "oil",
    "CESPM 1 FO": "oil",
    "CESPM 1 GN": "gas",
    "CESPM 2 FO": "oil",
    "CESPM 2 GN": "gas",
    "CESPM 3 FO": "oil",
    "CESPM 3 GN": "gas",
    "ESTRELLA DEL MAR 2 CFO": "oil",
    "ESTRELLA DEL MAR 2 CGN": "gas",
    "ESTRELLA DEL MAR 2 SFO": "oil",
    "ESTRELLA DEL MAR 2 SGN": "gas",
    "ESTRELLA DEL MAR 3": "gas",
    "GENERACI\xd3N DE EMERGENCIA AES ANDR\xc9S": "gas",
    "HAINA TG": "oil",
    "INCA KM22": "oil",
    "ITABO 1": "coal",
    "ITABO 2": "coal",
    "LA VEGA": "oil",
    "LOS MINA 5": "gas",
    "LOS MINA 6": "gas",
    "LOS MINA 7": "gas",
    "LOS OR\xcdGENES POWER PLANT FUEL OIL": "oil",
    "LOS OR\xcdGENES POWER PLANT GAS NATURAL": "gas",
    "METALDOM": "oil",
    "MONTE RIO": "oil",
    "PALAMARA": "oil",
    "PALENQUE": "oil",
    "PARQUE ENERGETICO LOS MINA CC PARCIAL": "gas",
    "PARQUE ENERGETICO LOS MINA CC TOTAL": "gas",
    "PIMENTEL 1": "oil",
    "PIMENTEL 2": "oil",
    "PIMENTEL 3": "oil",
    "PUNTA CATALINA 1": "coal",
    "PUNTA CATALINA 2": "coal",
    "QUISQUEYA 1B SAN PEDRO GN": "gas",
    "QUISQUEYA 1 FO": "oil",
    "QUISQUEYA 1 GN": "gas",
    "QUISQUEYA 2 FO": "oil",
    "QUISQUEYA 2 GN": "gas",
    "QUISQUEYA 1 SAN PEDRO FO": "oil",
    "QUISQUEYA 1 SAN PEDRO GN": "gas",
    "RIO SAN JUAN": "oil",
    "SAN FELIPE": "oil",
    "SAN FELIPE CC": "gas",
    "SAN FELIPE VAP": "oil",
    "SAN LORENZO 1": "gas",
    "SAN PEDRO BIO-ENERGY": "biomass",
    "SAN PEDRO VAPOR": "oil",
    "SULTANA DEL ESTE": "oil",
}


def get_datetime_from_hour(now: datetime, hour: int) -> datetime:
    return now + timedelta(hours=int(hour) - 1)


def get_data(session: Session | None = None) -> list[list[str]]:
    """
    Makes a request to source url.
    Finds main table and creates a list of all table elements in string format.
    """

    s = session or Session()
    data_req = s.get(URL)
    soup = BeautifulSoup(data_req.content, "lxml")

    tbs = soup.find("table", id="PostdespachoUnidadesTermicasGrid_DXMainTable")
    rows = tbs.find_all("tr")

    data = []
    for row in rows:
        row_data = []
        cols = row.find_all("td")
        for col in cols:
            row_data.append(str(col.getText().strip()))
        data.append(row_data)

    return data


def floater(item):
    """
    Attempts to convert any item given to a float.
    Returns item if it fails.
    """

    try:
        return float(item)
    except ValueError:
        return item


def chunker(big_lst) -> dict:
    """
    Breaks a big list into a list of lists.
    Removes any list with no data then turns remaining
    lists into key: value pairs with first element from the list being the key.
    """

    chunks = [big_lst[x : x + 27] for x in range(0, len(big_lst), 27)]

    # Remove the list if it contains no data.
    for chunk in chunks:
        if any(chunk):
            continue
        else:
            chunks.remove(chunk)

    chunked_list = {words[0]: words[1:] for words in chunks}

    return chunked_list


def data_formatter(data: list[list[str]]) -> list[list[str]]:
    """
    Aligns the tabular data to a standard format: (ID, hour_0, hour_1, ... , hour_23, hour_24)
    """

    INIT_ROWS_TO_DROP = 26
    data = data[INIT_ROWS_TO_DROP:]

    def format_row(row: list[str]) -> list[str]:
        # Case Grupo: X
        match_grupo = len(row) == 2 and row[0] == "" and "grupo" in row[1].lower()
        # Case Empresa: X
        match_empresa = (
            len(row) == 3
            and all(c == "" for c in row[:2])
            and "empresa" in row[2].lower()
        )
        # Case Unit: X
        match_unit = len(row) == 27 and all(c == "" for c in row[:2])

        if match_grupo:
            return [row[1]] + [""] * 24
        elif match_empresa:
            return [row[2]] + [""] * 24
        elif match_unit:
            return row[2:]
        else:
            raise ValueError(f"Unexpected row format: {row}")

    data = [format_row(row) for row in data]
    return data


def correct_solar_production(production: pd.DataFrame) -> pd.DataFrame:
    """
    Solar production is not reported when it's zero.
    """
    if production.solar.isnull().all() or production.solar.notnull().all():
        return production
    production = production.copy()
    null_production_index = production[production.solar.isnull()].index
    max_non_null_solar_idx = production.solar.last_valid_index()
    indices_to_set_to_zero = [
        idx for idx in null_production_index if idx < max_non_null_solar_idx
    ]
    # Replace all NaN values with 0 up to the first non-null value
    production.loc[indices_to_set_to_zero] = 0
    return production


def extract_renewable_production(data: list[list[str]], dt: datetime) -> pd.DataFrame:
    """
    Extract renewable production data from the total rows.
    """
    renewable_indices = [
        i for i, row in enumerate(data) if row[0] in TOTAL_RENEWABLES_MAPPING
    ]
    renewable_data = []
    for i in renewable_indices:
        row = data[i]
        renewable_data.append([TOTAL_RENEWABLES_MAPPING[row[0]]] + row[1:])
    df = pd.DataFrame(renewable_data, columns=["mode"] + list(range(1, 25)))
    # pivot to have hours as index and mode as columns
    df = df.set_index("mode").T
    df.index = [get_datetime_from_hour(dt, hour) for hour in df.index]
    df.index.name = "datetime"
    # Convert to numeric
    df = df.apply(pd.to_numeric)
    df = correct_solar_production(df)
    return df


def extract_thermal_production(data: list[list[str]], dt: datetime) -> pd.DataFrame:
    """
    Extract thermal production from individual power plants.
    """
    thermal_indices = [i for i, row in enumerate(data) if row[0] in THERMAL_PLANTS]
    thermal_data = []
    for i in thermal_indices:
        row = data[i]
        thermal_data.append([THERMAL_PLANTS.get(row[0], "unknown")] + row[1:])
    df = pd.DataFrame(thermal_data, columns=["mode"] + list(range(1, 25)))
    # Convert numeric
    df = df.apply(pd.to_numeric, errors="ignore")
    # Group by sum per mode
    df = df.groupby("mode").sum(min_count=1)
    # pivot to have hours as index and mode as columns
    df = df.T
    df.index = [get_datetime_from_hour(dt, hour) for hour in df.index]
    df.index.name = "datetime"
    return df


@use_proxy(country_code="DO")
def fetch_production(
    zone_key: ZoneKey = ZoneKey("DO"),
    session: Session | None = None,
    target_datetime: datetime | None = None,
    logger: Logger = getLogger(__name__),
) -> list[dict]:
    """Requests the last known production mix (in MW) of a given country."""
    if target_datetime:
        raise NotImplementedError("This parser is not yet able to parse past dates")

    now = datetime.now(tz=ZoneInfo("America/Dominica")).replace(
        hour=0, minute=0, second=0, microsecond=0
    )

    data = data_formatter(get_data(session=session))
    renewable_production = extract_renewable_production(data, now)
    thermal_production = extract_thermal_production(data, now)
    production = pd.concat([renewable_production, thermal_production], axis=1)
    # only keep rows with at least one non-null value
    production = production.dropna(how="all")

    production_list = ProductionBreakdownList(logger)
    for ts, mix in production.iterrows():
        production_mix = ProductionMix(**mix.to_dict())
        production_list.append(
            zoneKey=zone_key,
            datetime=ts.to_pydatetime(),
            source=DO_SOURCE,
            production=production_mix,
        )

    return production_list.to_list()


if __name__ == "__main__":
    """Main method, never used by the Electricity Map backend, but handy for testing."""

    print("fetch_production() ->")
    print(fetch_production())
