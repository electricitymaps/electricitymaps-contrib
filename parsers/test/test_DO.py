from importlib import resources

import freezegun
import pandas as pd
import pytest
from numpy import nan
from requests_mock import ANY

from electricitymap.contrib.parsers.DO import correct_solar_production, fetch_production


@pytest.fixture
def production_df():
    d = {
        "wind": {
            pd.Timestamp("2024-04-18 00:00:00-0400", tz="America/Dominica"): 117.42,
            pd.Timestamp("2024-04-18 01:00:00-0400", tz="America/Dominica"): 76.37,
            pd.Timestamp("2024-04-18 02:00:00-0400", tz="America/Dominica"): 53.67,
            pd.Timestamp("2024-04-18 03:00:00-0400", tz="America/Dominica"): 57.76,
            pd.Timestamp("2024-04-18 04:00:00-0400", tz="America/Dominica"): 64.95,
            pd.Timestamp("2024-04-18 05:00:00-0400", tz="America/Dominica"): 52.9,
            pd.Timestamp("2024-04-18 06:00:00-0400", tz="America/Dominica"): 46.25,
            pd.Timestamp("2024-04-18 07:00:00-0400", tz="America/Dominica"): nan,
            pd.Timestamp("2024-04-18 08:00:00-0400", tz="America/Dominica"): nan,
            pd.Timestamp("2024-04-18 09:00:00-0400", tz="America/Dominica"): nan,
            pd.Timestamp("2024-04-18 10:00:00-0400", tz="America/Dominica"): nan,
            pd.Timestamp("2024-04-18 11:00:00-0400", tz="America/Dominica"): nan,
            pd.Timestamp("2024-04-18 12:00:00-0400", tz="America/Dominica"): nan,
            pd.Timestamp("2024-04-18 13:00:00-0400", tz="America/Dominica"): nan,
            pd.Timestamp("2024-04-18 14:00:00-0400", tz="America/Dominica"): nan,
            pd.Timestamp("2024-04-18 15:00:00-0400", tz="America/Dominica"): nan,
            pd.Timestamp("2024-04-18 16:00:00-0400", tz="America/Dominica"): nan,
            pd.Timestamp("2024-04-18 17:00:00-0400", tz="America/Dominica"): nan,
            pd.Timestamp("2024-04-18 18:00:00-0400", tz="America/Dominica"): nan,
            pd.Timestamp("2024-04-18 19:00:00-0400", tz="America/Dominica"): nan,
            pd.Timestamp("2024-04-18 20:00:00-0400", tz="America/Dominica"): nan,
            pd.Timestamp("2024-04-18 21:00:00-0400", tz="America/Dominica"): nan,
            pd.Timestamp("2024-04-18 22:00:00-0400", tz="America/Dominica"): nan,
            pd.Timestamp("2024-04-18 23:00:00-0400", tz="America/Dominica"): nan,
        },
        "solar": {
            pd.Timestamp("2024-04-18 00:00:00-0400", tz="America/Dominica"): nan,
            pd.Timestamp("2024-04-18 01:00:00-0400", tz="America/Dominica"): nan,
            pd.Timestamp("2024-04-18 02:00:00-0400", tz="America/Dominica"): nan,
            pd.Timestamp("2024-04-18 03:00:00-0400", tz="America/Dominica"): nan,
            pd.Timestamp("2024-04-18 04:00:00-0400", tz="America/Dominica"): nan,
            pd.Timestamp("2024-04-18 05:00:00-0400", tz="America/Dominica"): nan,
            pd.Timestamp("2024-04-18 06:00:00-0400", tz="America/Dominica"): nan,
            pd.Timestamp("2024-04-18 07:00:00-0400", tz="America/Dominica"): nan,
            pd.Timestamp("2024-04-18 08:00:00-0400", tz="America/Dominica"): nan,
            pd.Timestamp("2024-04-18 09:00:00-0400", tz="America/Dominica"): nan,
            pd.Timestamp("2024-04-18 10:00:00-0400", tz="America/Dominica"): nan,
            pd.Timestamp("2024-04-18 11:00:00-0400", tz="America/Dominica"): nan,
            pd.Timestamp("2024-04-18 12:00:00-0400", tz="America/Dominica"): nan,
            pd.Timestamp("2024-04-18 13:00:00-0400", tz="America/Dominica"): nan,
            pd.Timestamp("2024-04-18 14:00:00-0400", tz="America/Dominica"): nan,
            pd.Timestamp("2024-04-18 15:00:00-0400", tz="America/Dominica"): nan,
            pd.Timestamp("2024-04-18 16:00:00-0400", tz="America/Dominica"): nan,
            pd.Timestamp("2024-04-18 17:00:00-0400", tz="America/Dominica"): nan,
            pd.Timestamp("2024-04-18 18:00:00-0400", tz="America/Dominica"): nan,
            pd.Timestamp("2024-04-18 19:00:00-0400", tz="America/Dominica"): nan,
            pd.Timestamp("2024-04-18 20:00:00-0400", tz="America/Dominica"): nan,
            pd.Timestamp("2024-04-18 21:00:00-0400", tz="America/Dominica"): nan,
            pd.Timestamp("2024-04-18 22:00:00-0400", tz="America/Dominica"): nan,
            pd.Timestamp("2024-04-18 23:00:00-0400", tz="America/Dominica"): nan,
        },
        "hydro": {
            pd.Timestamp("2024-04-18 00:00:00-0400", tz="America/Dominica"): 144.47,
            pd.Timestamp("2024-04-18 01:00:00-0400", tz="America/Dominica"): 75.34,
            pd.Timestamp("2024-04-18 02:00:00-0400", tz="America/Dominica"): 72.94,
            pd.Timestamp("2024-04-18 03:00:00-0400", tz="America/Dominica"): 84.23,
            pd.Timestamp("2024-04-18 04:00:00-0400", tz="America/Dominica"): 84.52,
            pd.Timestamp("2024-04-18 05:00:00-0400", tz="America/Dominica"): 84.68,
            pd.Timestamp("2024-04-18 06:00:00-0400", tz="America/Dominica"): 87.36,
            pd.Timestamp("2024-04-18 07:00:00-0400", tz="America/Dominica"): nan,
            pd.Timestamp("2024-04-18 08:00:00-0400", tz="America/Dominica"): nan,
            pd.Timestamp("2024-04-18 09:00:00-0400", tz="America/Dominica"): nan,
            pd.Timestamp("2024-04-18 10:00:00-0400", tz="America/Dominica"): nan,
            pd.Timestamp("2024-04-18 11:00:00-0400", tz="America/Dominica"): nan,
            pd.Timestamp("2024-04-18 12:00:00-0400", tz="America/Dominica"): nan,
            pd.Timestamp("2024-04-18 13:00:00-0400", tz="America/Dominica"): nan,
            pd.Timestamp("2024-04-18 14:00:00-0400", tz="America/Dominica"): nan,
            pd.Timestamp("2024-04-18 15:00:00-0400", tz="America/Dominica"): nan,
            pd.Timestamp("2024-04-18 16:00:00-0400", tz="America/Dominica"): nan,
            pd.Timestamp("2024-04-18 17:00:00-0400", tz="America/Dominica"): nan,
            pd.Timestamp("2024-04-18 18:00:00-0400", tz="America/Dominica"): nan,
            pd.Timestamp("2024-04-18 19:00:00-0400", tz="America/Dominica"): nan,
            pd.Timestamp("2024-04-18 20:00:00-0400", tz="America/Dominica"): nan,
            pd.Timestamp("2024-04-18 21:00:00-0400", tz="America/Dominica"): nan,
            pd.Timestamp("2024-04-18 22:00:00-0400", tz="America/Dominica"): nan,
            pd.Timestamp("2024-04-18 23:00:00-0400", tz="America/Dominica"): nan,
        },
    }
    return pd.DataFrame(d)


def test_correct_solar_production_all_nan(production_df):
    corrected_df = correct_solar_production(production_df)
    assert corrected_df["solar"].isna().all()


def test_correct_solar_production_nan_then_prod(production_df):
    production_df["solar"].loc[
        pd.Timestamp("2024-04-18 06:00:00-0400", tz="America/Dominica")
    ] = 12
    corrected_df = correct_solar_production(production_df)
    assert (
        corrected_df["solar"].loc[
            pd.Timestamp("2024-04-18 06:00:00-0400", tz="America/Dominica")
        ]
        == 12
    )
    assert all(
        corrected_df["solar"].loc[
            : pd.Timestamp("2024-04-18 05:00:00-0400", tz="America/Dominica")
        ]
        == 0
    )
    assert all(
        corrected_df["solar"]
        .loc[pd.Timestamp("2024-04-18 07:00:00-0400", tz="America/Dominica") :]
        .isnull()
    )


def test_correct_solar_production_prod_then_nan(production_df):
    production_df["solar"].loc[
        : pd.Timestamp("2024-04-18 03:00:00-0400", tz="America/Dominica")
    ] = 12
    corrected_df = correct_solar_production(production_df)
    assert all(
        corrected_df["solar"].loc[
            : pd.Timestamp("2024-04-18 03:00:00-0400", tz="America/Dominica")
        ]
        == 12
    )


def test_correct_solar_production_prod_then_nan_then_prod(production_df):
    production_df["solar"].loc[
        : pd.Timestamp("2024-04-18 03:00:00-0400", tz="America/Dominica")
    ] = 12
    production_df["solar"].loc[
        pd.Timestamp("2024-04-18 06:00:00-0400", tz="America/Dominica")
    ] = 14
    corrected_df = correct_solar_production(production_df)
    assert all(
        corrected_df["solar"].loc[
            : pd.Timestamp("2024-04-18 03:00:00-0400", tz="America/Dominica")
        ]
        == 12
    )
    assert (
        corrected_df["solar"].loc[
            pd.Timestamp("2024-04-18 06:00:00-0400", tz="America/Dominica")
        ]
        == 14
    )
    assert all(
        corrected_df["solar"].loc[
            pd.Timestamp(
                "2024-04-18 04:00:00-0400", tz="America/Dominica"
            ) : pd.Timestamp("2024-04-18 05:00:00-0400", tz="America/Dominica")
        ]
        == 0
    )
    assert all(
        corrected_df["solar"]
        .loc[pd.Timestamp("2024-04-18 07:00:00-0400", tz="America/Dominica") :]
        .isnull()
    )


@freezegun.freeze_time("2025-01-15")
def test_fetch_production(adapter, session, snapshot):
    adapter.register_uri(
        ANY,
        ANY,
        content=resources.files("parsers.test.mocks.DO")
        .joinpath("production.html")
        .read_bytes(),
    )

    assert snapshot == fetch_production(session=session)
