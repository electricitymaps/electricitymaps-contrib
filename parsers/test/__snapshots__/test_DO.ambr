# serializer version: 1
# name: test_fetch_production
  list([
    dict({
      'correctedModes': list([
      ]),
      'datetime': datetime.datetime(2025, 1, 14, 0, 0, tzinfo=zoneinfo.ZoneInfo(key='America/Dominica')),
      'production': dict({
        'biomass': 30.02,
        'coal': 940.42,
        'gas': 914.61,
        'hydro': 0.0,
        'oil': 171.16,
        'solar': 0.0,
        'wind': 0.0,
      }),
      'source': 'oc.org.do',
      'sourceType': <EventSourceType.measured: 'measured'>,
      'storage': dict({
      }),
      'zoneKey': 'DO',
    }),
    dict({
      'correctedModes': list([
      ]),
      'datetime': datetime.datetime(2025, 1, 14, 1, 0, tzinfo=zoneinfo.ZoneInfo(key='America/Dominica')),
      'production': dict({
        'biomass': 30.05,
        'coal': 937.01,
        'gas': 918.43,
        'hydro': 0.0,
        'oil': 198.85,
        'solar': 0.0,
        'wind': 0.0,
      }),
      'source': 'oc.org.do',
      'sourceType': <EventSourceType.measured: 'measured'>,
      'storage': dict({
      }),
      'zoneKey': 'DO',
    }),
    dict({
      'correctedModes': list([
      ]),
      'datetime': datetime.datetime(2025, 1, 14, 2, 0, tzinfo=zoneinfo.ZoneInfo(key='America/Dominica')),
      'production': dict({
        'biomass': 30.1,
        'coal': 936.61,
        'gas': 867.07,
        'hydro': 0.0,
        'oil': 262.37,
        'solar': 0.0,
        'wind': 0.0,
      }),
      'source': 'oc.org.do',
      'sourceType': <EventSourceType.measured: 'measured'>,
      'storage': dict({
      }),
      'zoneKey': 'DO',
    }),
    dict({
      'correctedModes': list([
      ]),
      'datetime': datetime.datetime(2025, 1, 14, 3, 0, tzinfo=zoneinfo.ZoneInfo(key='America/Dominica')),
      'production': dict({
        'biomass': 30.1,
        'coal': 936.55,
        'gas': 806.58,
        'hydro': 0.0,
        'oil': 264.72,
        'solar': 0.0,
        'wind': 0.0,
      }),
      'source': 'oc.org.do',
      'sourceType': <EventSourceType.measured: 'measured'>,
      'storage': dict({
      }),
      'zoneKey': 'DO',
    }),
    dict({
      'correctedModes': list([
      ]),
      'datetime': datetime.datetime(2025, 1, 14, 4, 0, tzinfo=zoneinfo.ZoneInfo(key='America/Dominica')),
      'production': dict({
        'biomass': 29.72,
        'coal': 932.76,
        'gas': 827.35,
        'hydro': 0.0,
        'oil': 177.87,
        'solar': 0.0,
        'wind': 0.0,
      }),
      'source': 'oc.org.do',
      'sourceType': <EventSourceType.measured: 'measured'>,
      'storage': dict({
      }),
      'zoneKey': 'DO',
    }),
    dict({
      'correctedModes': list([
      ]),
      'datetime': datetime.datetime(2025, 1, 14, 5, 0, tzinfo=zoneinfo.ZoneInfo(key='America/Dominica')),
      'production': dict({
        'biomass': 30.06,
        'coal': 928.92,
        'gas': 902.21,
        'hydro': 0.0,
        'oil': 119.51,
        'solar': 0.0,
        'wind': 0.0,
      }),
      'source': 'oc.org.do',
      'sourceType': <EventSourceType.measured: 'measured'>,
      'storage': dict({
      }),
      'zoneKey': 'DO',
    }),
    dict({
      'correctedModes': list([
      ]),
      'datetime': datetime.datetime(2025, 1, 14, 6, 0, tzinfo=zoneinfo.ZoneInfo(key='America/Dominica')),
      'production': dict({
        'biomass': 28.46,
        'coal': 930.8,
        'gas': 928.82,
        'hydro': 0.0,
        'oil': 119.18,
        'solar': 0.0,
        'wind': 0.0,
      }),
      'source': 'oc.org.do',
      'sourceType': <EventSourceType.measured: 'measured'>,
      'storage': dict({
      }),
      'zoneKey': 'DO',
    }),
    dict({
      'correctedModes': list([
      ]),
      'datetime': datetime.datetime(2025, 1, 14, 7, 0, tzinfo=zoneinfo.ZoneInfo(key='America/Dominica')),
      'production': dict({
        'biomass': 28.87,
        'coal': 924.34,
        'gas': 880.56,
        'hydro': 114.06,
        'oil': 126.49,
        'solar': 23.9,
        'wind': 9.91,
      }),
      'source': 'oc.org.do',
      'sourceType': <EventSourceType.measured: 'measured'>,
      'storage': dict({
      }),
      'zoneKey': 'DO',
    }),
    dict({
      'correctedModes': list([
      ]),
      'datetime': datetime.datetime(2025, 1, 14, 8, 0, tzinfo=zoneinfo.ZoneInfo(key='America/Dominica')),
      'production': dict({
        'biomass': 27.89,
        'coal': 924.49,
        'gas': 772.35,
        'hydro': 72.41,
        'oil': 73.84,
        'solar': 251.22,
        'wind': 8.68,
      }),
      'source': 'oc.org.do',
      'sourceType': <EventSourceType.measured: 'measured'>,
      'storage': dict({
      }),
      'zoneKey': 'DO',
    }),
    dict({
      'correctedModes': list([
      ]),
      'datetime': datetime.datetime(2025, 1, 14, 9, 0, tzinfo=zoneinfo.ZoneInfo(key='America/Dominica')),
      'production': dict({
        'biomass': 25.67,
        'coal': 926.15,
        'gas': 615.72,
        'hydro': 67.23,
        'oil': None,
        'solar': 527.22,
        'wind': 5.94,
      }),
      'source': 'oc.org.do',
      'sourceType': <EventSourceType.measured: 'measured'>,
      'storage': dict({
      }),
      'zoneKey': 'DO',
    }),
    dict({
      'correctedModes': list([
      ]),
      'datetime': datetime.datetime(2025, 1, 14, 10, 0, tzinfo=zoneinfo.ZoneInfo(key='America/Dominica')),
      'production': dict({
        'biomass': 27.27,
        'coal': 868.14,
        'gas': 549.8,
        'hydro': 65.77,
        'oil': None,
        'solar': 612.36,
        'wind': 8.79,
      }),
      'source': 'oc.org.do',
      'sourceType': <EventSourceType.measured: 'measured'>,
      'storage': dict({
      }),
      'zoneKey': 'DO',
    }),
    dict({
      'correctedModes': list([
      ]),
      'datetime': datetime.datetime(2025, 1, 14, 11, 0, tzinfo=zoneinfo.ZoneInfo(key='America/Dominica')),
      'production': dict({
        'biomass': 29.75,
        'coal': 827.64,
        'gas': 576.89,
        'hydro': 58.3,
        'oil': None,
        'solar': 653.86,
        'wind': 16.17,
      }),
      'source': 'oc.org.do',
      'sourceType': <EventSourceType.measured: 'measured'>,
      'storage': dict({
      }),
      'zoneKey': 'DO',
    }),
    dict({
      'correctedModes': list([
      ]),
      'datetime': datetime.datetime(2025, 1, 14, 12, 0, tzinfo=zoneinfo.ZoneInfo(key='America/Dominica')),
      'production': dict({
        'biomass': 29.78,
        'coal': 820.55,
        'gas': 577.44,
        'hydro': 54.27,
        'oil': None,
        'solar': 679.36,
        'wind': 33.4,
      }),
      'source': 'oc.org.do',
      'sourceType': <EventSourceType.measured: 'measured'>,
      'storage': dict({
      }),
      'zoneKey': 'DO',
    }),
    dict({
      'correctedModes': list([
      ]),
      'datetime': datetime.datetime(2025, 1, 14, 13, 0, tzinfo=zoneinfo.ZoneInfo(key='America/Dominica')),
      'production': dict({
        'biomass': 29.91,
        'coal': 825.69,
        'gas': 578.61,
        'hydro': 42.51,
        'oil': None,
        'solar': 695.16,
        'wind': 49.56,
      }),
      'source': 'oc.org.do',
      'sourceType': <EventSourceType.measured: 'measured'>,
      'storage': dict({
      }),
      'zoneKey': 'DO',
    }),
    dict({
      'correctedModes': list([
      ]),
      'datetime': datetime.datetime(2025, 1, 14, 14, 0, tzinfo=zoneinfo.ZoneInfo(key='America/Dominica')),
      'production': dict({
        'biomass': 30.0,
        'coal': 860.69,
        'gas': 645.81,
        'hydro': 42.39,
        'oil': None,
        'solar': 676.98,
        'wind': 74.53,
      }),
      'source': 'oc.org.do',
      'sourceType': <EventSourceType.measured: 'measured'>,
      'storage': dict({
      }),
      'zoneKey': 'DO',
    }),
    dict({
      'correctedModes': list([
      ]),
      'datetime': datetime.datetime(2025, 1, 14, 15, 0, tzinfo=zoneinfo.ZoneInfo(key='America/Dominica')),
      'production': dict({
        'biomass': 28.99,
        'coal': 923.92,
        'gas': 650.54,
        'hydro': 42.53,
        'oil': 2.09,
        'solar': 599.58,
        'wind': 83.42,
      }),
      'source': 'oc.org.do',
      'sourceType': <EventSourceType.measured: 'measured'>,
      'storage': dict({
      }),
      'zoneKey': 'DO',
    }),
    dict({
      'correctedModes': list([
      ]),
      'datetime': datetime.datetime(2025, 1, 14, 16, 0, tzinfo=zoneinfo.ZoneInfo(key='America/Dominica')),
      'production': dict({
        'biomass': 29.19,
        'coal': 933.95,
        'gas': 794.39,
        'hydro': 78.93,
        'oil': 42.77,
        'solar': 446.65,
        'wind': 75.95,
      }),
      'source': 'oc.org.do',
      'sourceType': <EventSourceType.measured: 'measured'>,
      'storage': dict({
      }),
      'zoneKey': 'DO',
    }),
    dict({
      'correctedModes': list([
      ]),
      'datetime': datetime.datetime(2025, 1, 14, 17, 0, tzinfo=zoneinfo.ZoneInfo(key='America/Dominica')),
      'production': dict({
        'biomass': 30.09,
        'coal': 932.86,
        'gas': 810.65,
        'hydro': 128.41,
        'oil': 100.27,
        'solar': 279.41,
        'wind': 63.48,
      }),
      'source': 'oc.org.do',
      'sourceType': <EventSourceType.measured: 'measured'>,
      'storage': dict({
      }),
      'zoneKey': 'DO',
    }),
  ])
# ---
